---
title: "DE-iCluster"
author: "Ruqian"
date: "04/04/2018"
  
bibliography: ./ref.bib
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction

This documents explains the steps taken for deriving DE gene sets for 3 iClusters of HCC patient samples.
The datasets were published by https://doi.org/10.1016/j.cell.2017.05.046 

## Data packaging

### Read in count-data

The raw count data were downloaded from the Genomic Data Commons Data Portal https://portal.gdc.cancer.gov/ from project TCGA-LIHC. 424 samples were downloaded using the following filter rules.![](./filefilter.png)

The study by [@cihcc] had clustered the HCC patient samples into three groups. 
The sample IDs of patients within each cluster can be found in the TableS1.


```{r readinmmc1}
library(readxl)
mmc1 <- read_excel("~/Documents/davisLab/mmc1.xlsx", skip = 3)
head(mmc1$`iCluster clusters (k=3, Ronglai Shen)`)
```

We will analyse samples that have been clustered, so we remoe the "NA" rows.

```{r removeNA}

identified_clusters <- mmc1[! mmc1$`iCluster clusters (k=3, Ronglai Shen)`=="NA",]

# So we will be looking at 183 samples
dim(identified_clusters)

```

### Load the raw read count data 
Next we load the raw read count data from our downloaded files.

```{r sampleIDMapping}
library(edgeR)
# get the sample IDs of 183 samples
clustered_samples <- identified_clusters$Barcode

# From the sample sheet downloaded from GDC, we get the file names mapped to the 
# SampleID
gdc_sample_sheet.2018.03.28 <- read.delim("~/Documents/davisLab/HCC/gdc_sample_sheet.2018-03-28.tsv", quote="")

all_files <- gdc_sample_sheet.2018.03.28[gdc_sample_sheet.2018.03.28$Sample.ID %in% clustered_samples,]

dim(all_files)
```
Load the raw count data from files

```{r loadCountFiles}

files_path <- paste0("/Users/ruqianlyu/Documents/davisLab/HCC/gdc_download_20180328_001539/",all_files$File.ID, "/", all_files$File.Name)

# Use the readDGE function for edgeR for reading all samples' count files.
initial_dge_object <- edgeR::readDGE(files_path)

# Here is the dimension of the raw gene expression dataset
dim(initial_dge_object)

# Use shorter file names 
files_names <- substring(colnames(initial_dge_object), 107, nchar(colnames(initial_dge_object)))

head(files_names)

# rename the column
colnames(initial_dge_object) <- files_names 

# Map the file_names to Sample.ID so that we can identify the cluster of the counts
y <- as.data.frame(colnames(initial_dge_object))
colnames(y) <- "File.Name"

k <- as.data.frame(paste0(y$File.Name,".gz"))

colnames(k) <- "File.Name"

head(k)

# Using the sample sheet document to align the file name with Sample.ID
f <- merge(k, gdc_sample_sheet.2018.03.28, by = "File.Name", sort = FALSE)

head(f)

# Use the mmc1 file records to annotate the Samples with cluster labels
g <- merge(f, mmc1, by.x = "Sample.ID", by.y = "Barcode", sort = FALSE)

head(g)

# The factors that might be of interest
race <- as.factor(g$race)
gender <- as.factor(g$gender)

# group stores the cluster label for each sample
group <- as.factor(g$`iCluster clusters (k=3, Ronglai Shen)`)

#remove the ":" from cluster label
temp <- sapply(group, function(x){sub(":","",x)})
group <- temp

initial_dge_object$samples$group <- group
initial_dge_object$samples$race <- race
initial_dge_object$samples$gender <- gender

# Here we have the initial_dge_object which contains the count matrxi and sample
# information for 183 samples and 60487 ESEMBL transcript IDs
dim(initial_dge_object)
```

## Mapping gene annotation and filtering

## Check gene types with `bioMart`

`bioMart` is an R pacakge that provides the interface with ???. Thus `bioMart` can be 
used to ??, we will only keep the protein coding genes in our analysis.

```{r bioMartGeneType, eval=FALSE}
library(biomaRt)
ensembl=useMart("ensembl")
#listDatasets(ensembl)
ensembl = useDataset("hsapiens_gene_ensembl",mart=ensembl)
#listFilters(ensembl)
attrs = listAttributes(ensembl)

# mapToType = getBM(
#   attributes = c('transcript_biotype', 'entrezgene','ensembl_gene_id',
# 'ensembl_gene_id_version','hgnc_symbol'),
#   filters = 'entrezgene',
#   values = genes$ENTREZID,
#   mart = ensembl
# )
# save(mapToType,file = "mapToType.RData")

load("mapToType.RData")

head(mapToType)


# Find the mappings for transcripts in our intial dge object
mappedTrans = mapToType[mapToType$ensembl_gene_id %in% rownames(initial_dge_object$counts),]

# Find the proten coding genes
mappedTrans = mappedTrans[mappedTrans$transcript_biotype == "protein_coding",]

dim(mappedTrans)

# remove the duplicated mapping by symbol
dup <- mappedTrans$hgnc_symbol[duplicated(mappedTrans$hgnc_symbol)]
length(dup)

# The first 5 duplicated ones are 
mappedTrans[mappedTrans$hgnc_symbol %in% dup,][1:5,]

# takes care of the duplication by keeping the first occurrence of each gene ID
# match returns a vector of the positions of (first) matches of its first argument in its 
# second
mat <- match(unique(mappedTrans$hgnc_symbol), mappedTrans$hgnc_symbol)
mappedTrans <- mappedTrans[mat,]

dim(mappedTrans)

# Use the `ENSEMBL` column remained in the `genes` to subset the initial dge object 
dge_object <- initial_dge_object[mappedTrans$ensembl_gene_id,]
dge_object$genes <- mappedTrans
rownames(dge_object) = mappedTrans$hgnc_symbol
# Cleaned dge_object
dim(dge_object)



```


## Data pre-processing for deriving DE genes among three iClusters

### Transformations from the raw-scale

```{r normalisation}

# raw counts are converted to CPM and log-CPM values using the CPM function 
# in edgeR

cpmVal <- cpm(dge_object)
lcpm <- cpm(dge_object, log=TRUE)

# removing lowly expressed genes, 0 across all samples
table(rowSums(dge_object$counts==0)==183)

# keep genes expressed in at least 55 samples, 55 is the number of samples in the smallest group
keep.exprs <- rowSums(cpmVal>1)>=55

sum(keep.exprs)

dge_object <- dge_object[keep.exprs,, keep.lib.sizes=FALSE]
dim(dge_object)

```
Plot before and after

```{r plotCompare}
library(RColorBrewer)
nsamples <- ncol(dge_object)
col <- brewer.pal(nsamples, "Paired")
par(mfrow=c(1,2))
plot(density(lcpm[,1]), col=col[1], lwd=2, ylim=c(0,0.21), las=2, main="", xlab="")
title(main="A. Raw data", xlab="Log-cpm")
abline(v=0, lty=3)
for (i in 2:nsamples){
 den <- density(lcpm[,i])
}
lines(den$x, den$y, col=col[i], lwd=2)
# legend("topright", samplenames, text.col=col, bty="n")
lcpm <- cpm(dge_object, log=TRUE)
plot(density(lcpm[,1]), col=col[1], lwd=2, ylim=c(0,0.21), las=2,
main="", xlab="")
title(main="B. Filtered data", xlab="Log-cpm") 
abline(v=0, lty=3)
for (i in 2:nsamples){
   den <- density(lcpm[,i])
}  
lines(den$x, den$y, col=col[i], lwd=2)
# legend("topright", samplenames, text.col=col, bty="n")
```

Normalisation by the method of trimmed mean of M-values (TMM)

```{r TMM}
dge_object <- calcNormFactors(dge_object, method = "TMM") 
head(dge_object$samples$norm.factors)

```
## MDS plot

To create the MDS plots, different colour groupings are assigned to factors of interest. Dimensions 1 and 2 are examined using the color grouping de ned by iClusters.

```{r MDS}
library(RColorBrewer)
lcpm <- cpm(dge_object, log=TRUE)
par(mfrow=c(1,2))
col.group <- group
col.group <- as.factor(col.group)
levels(col.group) <- brewer.pal(nlevels(col.group), "Set1") 
col.group <- as.character(col.group)
plotMDS(lcpm, pch = 19, col=col.group) 
title(main="A. Sample groups")

# The clusters do not look very promising, probably due to the clusters are formed not only by
# the transcriptome data unlike the original paper ?

# 1 separates from 2 and 3

# 2 and 3 more close to each other

# plotMDS(lcpm, labels=race, col=col.race, dim=c(3,4)) 
# title(main="B. Races")

## Maybe try PCA
```

## Differential expression analysis 

Pair-wise

```{r DEA}

design <- model.matrix(~0+group)

colnames(design) <- gsub("group", "", colnames(design)) 

contr.matrix <- makeContrasts(
   c1vsc2 = iCluster1-iCluster2,
   c2vsc3 = iCluster2-iCluster3,
   c1vsc3 = iCluster1-iCluster3,
   levels = colnames(design))
```

## Removing heteroscedasticity from count data

```{r}
v <- voom(dge_object, design, plot=TRUE)
# fit linear model
vfit <- lmFit(v, design)
vfit <- contrasts.fit(vfit, contrasts = contr.matrix)
efit <- eBayes(vfit)
plotSA(efit)

summary(decideTests(efit))

# A question here, do we need to use the same lfc threshold for 3 groups?
#tfit <- treat(vfit, lfc=0.32)
#dt <- decideTests(tfit)
#summary(dt)


c1.vs.c2 <- topTreat(efit, coef=1, n=Inf, p.value=0.01)
dim(c1.vs.c2)
c2.vs.c3 <- topTreat(efit, coef=2, n=Inf, p.value=0.01)
dim(c2.vs.c3)
c1.vs.c3 <- topTreat(efit, coef=3, n=Inf, p.value=0.01)
dim(c1.vs.c3)

# intersect the c1.vs.c2, c1.vs.c3
length(intersect(c1.vs.c2$hgnc_symbol, c1.vs.c3$hgnc_symbol))
length(intersect(c1.vs.c2$hgnc_symbol, c2.vs.c3$hgnc_symbol))
length(intersect(c1.vs.c3$hgnc_symbol, c2.vs.c3$hgnc_symbol))

sig1 = intersect(c1.vs.c2$hgnc_symbol, c1.vs.c3$hgnc_symbol)
sig2 = intersect(c1.vs.c2$hgnc_symbol, c2.vs.c3$hgnc_symbol)
sig3 = intersect(c1.vs.c3$hgnc_symbol, c2.vs.c3$hgnc_symbol)

sig1Df = data.frame(sig1 = sig1, p.1 = c1.vs.c2[sig1,"adj.P.Val"],logFC.1 = c1.vs.c2[sig1,"logFC"], p.2 = c1.vs.c3[sig1,"adj.P.Val"], logFC.2 = c1.vs.c3[sig1,"logFC"])
sig1Df$ps = sig1Df$p.1 + sig1Df$p.2

sig1_200 = head(sig1Df[order(sig1Df$ps),],200)$sig1

write.table(c1.vs.c2, file = "c1.vs.c2.txt")
write.table(c2.vs.c3, file = "c2.vs.c3.txt")
write.table(c1.vs.c3, file = "c1.vs.c3.txt")

#plotMD(tfit, column=1, status=dt[,1], main=colnames(tfit)[1], xlim=c(-8,13))
#plotMD(tfit, column=2, status=dt[,2], main=colnames(tfit)[2], xlim=c(-8,13))
#plotMD(tfit, column=3, status=dt[,3], main=colnames(tfit)[3], xlim=c(-8,13))

```
Could we only analyse the iCluster1 group ??, since it separates well from the remaining clusters and they are identified to have the lowest survical rate, therefore, when we score new samples, the new samples with high cluster1 gene signature concordance will be expected to have low survival rate.


```{r glimmaPlot,eval=FALSE}
library(Glimma)
lcpm <- cpm(dge_object, log = TRUE, normalized.lib.sizes = TRUE)
glMDPlot(tfit, coef=1, status=dt[,1], main=colnames(tfit)[1], counts=lcpm, samples=colnames(dge_object), anno=dge_object$genes, groups=group, side.main="ENTREZID", display.columns=c("SYMBOL", "ENTREZID"), search.by="SYMBOL", launch=TRUE)

glMDPlot(tfit, coef=2, status=dt[,2], main=colnames(tfit)[2], counts=dge_object$counts, samples=colnames(dge_object), anno=dge_object$genes, groups=group, side.main="ENTREZID", display.columns=c("SYMBOL", "ENTREZID"), search.by="SYMBOL", launch=FALSE)

```



```{r,eval=FALSE}
# Volcanoplot for the first contrast group
volcanoplot(tfit,coef=1,highlight=10)

# Volcanoplot for the second contrast group
volcanoplot(tfit,coef=2,highlight=10)

# Volcanoplot for the third contrast group
volcanoplot(tfit,coef=3,highlight=10)

```


Try different way of making contrast, this time we do 1 vs. rest

```{r contrast2}

design <- model.matrix(~0+group)
colnames(design) <- gsub("group", "", colnames(design)) 

contr.matrix2 <- makeContrasts(
   c1vsc2_3 = iCluster1-(iCluster2+iCluster3)/2,
   c2vsc1_3 = iCluster2-(iCluster1+iCluster3)/2,
   c3vsc1_2 = iCluster3-(iCluster1+iCluster2)/2,
   levels = colnames(design))
v <- voom(dge_object, design, plot=TRUE)
# remove the 

vfit <- lmFit(v, design)
vfit <- contrasts.fit(vfit, contrasts=contr.matrix2)
efit <- eBayes(vfit)
plotSA(efit)

summary(decideTests(efit))


tfit <- treat(efit, lfc=log2(1.27))
dt <- decideTests(tfit)
summary(dt)


new.c1.vs.c23 <- topTreat(tfit, coef=1, n=Inf, p.value = 0.01)
new.c2.vs.c13 <- topTreat(tfit, coef=2, n=Inf, p.value = 0.01, sort.by = 'p')
new.c3.vs.c12 <- topTreat(tfit, coef=3, n=Inf, p.value = 0.01, sort.by = 'p')

dim(new.c1.vs.c23)
dim(new.c2.vs.c13)
dim(new.c3.vs.c12)

write.table(new.c1.vs.c23, file = "new.c1.vs.c23.txt")
write.table(new.c2.vs.c13, file = "new.c2.vs.c13.txt")
write.table(new.c3.vs.c12, file = "new.c3.vs.c12.txt")

# Check overlap between the pair-wise comparision and non-pair-wise comparision
# sum(c1.vs.c2$ensembl_gene_id %in% new.c1.vs.c23$ensembl_gene_id)
# sum(c1.vs.c3$ensembl_gene_id %in% new.c1.vs.c23$ensembl_gene_id)
# 
# u1vs23 = union(c1.vs.c2$ensembl_gene_id, c1.vs.c3$ensembl_gene_id)
# length(u1vs23)
# sum(u1vs23 %in% new.c1.vs.c23$ensembl_gene_id)
# 
# sum(c2.vs.c3$ensembl_gene_id %in% new.c2.vs.c13$ensembl_gene_id)
# sum(c1.vs.c2$ensembl_gene_id %in% new.c2.vs.c13$ensembl_gene_id)
# 
# u2vs13 = union(c1.vs.c2$ensembl_gene_id, c2.vs.c3$ensembl_gene_id)
# length(u2vs13)
# sum(u2vs13 %in% new.c2.vs.c13$ensembl_gene_id)
# 
# sum(c1.vs.c3$ensembl_gene_id %in% new.c3.vs.c12$ensembl_gene_id)
# sum(c2.vs.c3$ensembl_gene_id %in% new.c3.vs.c12$ensembl_gene_id)
# 
# u3vs12 = union(c1.vs.c3$ensembl_gene_id, c2.vs.c3$ensembl_gene_id)
# length(u3vs12)
# sum(u3vs12 %in% new.c3.vs.c12$ensembl_gene_id)

# If we take a union of the pair-wise comparision result, and check whether those
# genes have been picked up in the non-pair wise result 
# The majory are in the non-pair-wise result

# u1vs23 
# 2269 / 2406

# u2vs13


glMDPlot(tfit, coef=1, status=dt[,1], main=colnames(tfit)[1], counts=dge_object$counts, samples=colnames(dge_object), anno=dge_object$genes, groups=group, side.main="ENTREZID", display.columns=c("SYMBOL", "ENTREZID"), search.by="SYMBOL", launch=FALSE)

plotMD(tfit, column=1, status=dt[,1], main=colnames(tfit)[1],xlim=c(-8,13))
plotMD(tfit, column=2, status=dt[,2], main=colnames(tfit)[2],xlim=c(-8,13))
plotMD(tfit, column=3, status=dt[,3], main=colnames(tfit)[3],xlim=c(-8,13))

volcanoplot(tfit,coef=1,highlight=10)
#volcanoplot(tfit,coef=2,highlight=10)
#volcanoplot(tfit,coef=3,highlight=10)
```

## Use the non-pair-wise result to score gene set

```{r summarizeGeneSet}
up_ic1 <- new.c1.vs.c23[new.c1.vs.c23$logFC>0,]
dn_ic1 <- new.c1.vs.c23[new.c1.vs.c23$logFC<0,]

up_ic2 <- new.c2.vs.c13[new.c2.vs.c13$logFC>0,]
dn_ic2 <- new.c2.vs.c13[new.c2.vs.c13$logFC<0,]

up_ic3 <- new.c3.vs.c12[new.c3.vs.c12$logFC>0,]
dn_ic3 <- new.c3.vs.c12[new.c3.vs.c12$logFC<0,]

up_ic1_gs <- up_ic1$hgnc_symbol
dn_ic1_gs <- dn_ic1$hgnc_symbol

up_ic2_gs <- up_ic2$hgnc_symbol
dn_ic2_gs <- dn_ic2$hgnc_symbol

up_ic3_gs <- up_ic3$hgnc_symbol
dn_ic3_gs <- dn_ic3$hgnc_symbol

```

```{r scoring}
library(singscore)
rankedData = rankGenes(dge_object$counts)
#rownames(rankedData) = dge_object$genes$hgnc_symbol

head(rankedData)
scoredf1 = simpleScore(rankedData, upSet = up_ic1_gs, downSet = dn_ic1_gs, centerScore = TRUE)
scoredf2 = simpleScore(rankedData, upSet = up_ic2_gs, downSet = dn_ic2_gs,centerScore = TRUE)
scoredf3 = simpleScore(rankedData, upSet = up_ic3_gs, downSet = dn_ic3_gs, centerScore = TRUE)

scoredf1 = simpleScore(rankedData, upSet = sig1_200, centerScore = TRUE, knownDirection = FALSE)
scoredf2 = simpleScore(rankedData, upSet = sig2,centerScore = TRUE,knownDirection = FALSE)
scoredf3 = simpleScore(rankedData, upSet = sig3,  centerScore = TRUE,knownDirection = FALSE)


plt1 = plotScoreLandscape(scoredf1,scoredf2, scorenames = c("IC1","IC2"))
plt2 = plotScoreLandscape(scoredf1,scoredf3, scorenames = c("IC1","IC3"))
plt3 = plotScoreLandscape(scoredf2,scoredf3, scorenames = c("IC2","IC3"))
#Overlay the iCluster1 on top of the landscape
ic1 = dge_object$counts[,dge_object$samples$group == "iCluster1"]
ic3 = dge_object$counts[,dge_object$samples$group == "iCluster3"]

rankedIC1 = rankGenes(ic1)
rankedIC3 = rankGenes(ic3)

scoredf_ic1 = simpleScore(rankedIC1, upSet = up_ic1_gs, downSet = dn_ic1_gs, centerScore = TRUE)
scoredf_ic2 = simpleScore(rankedIC1, upSet = up_ic2_gs, downSet = dn_ic2_gs,centerScore = TRUE)
scoredf_ic3 = simpleScore(rankedIC1, upSet = up_ic3_gs, downSet = dn_ic3_gs, centerScore = TRUE)

scoredf_ic1 = simpleScore(rankedIC1, upSet = sig1_200, centerScore = TRUE, knownDirection = FALSE)
scoredf_ic2 = simpleScore(rankedIC1, upSet = sig2, centerScore = TRUE,knownDirection = FALSE)
scoredf_ic3 = simpleScore(rankedIC1, upSet = sig3, centerScore = TRUE, knownDirection = FALSE)

scoredf_ic1_3 = simpleScore(rankedIC3, upSet = up_ic1_gs, downSet = dn_ic1_gs, centerScore = TRUE)
scoredf_ic2_3 = simpleScore(rankedIC3, upSet = up_ic2_gs, downSet = dn_ic2_gs,centerScore = TRUE)
scoredf_ic3_3 = simpleScore(rankedIC3, upSet = up_ic3_gs, downSet = dn_ic3_gs, centerScore = TRUE)

scoredf_ic1_3 = simpleScore(rankedIC3,upSet = sig1_200, centerScore = TRUE, knownDirection = FALSE)
scoredf_ic2_3 = simpleScore(rankedIC3, upSet = sig2, centerScore = TRUE,knownDirection = FALSE)
scoredf_ic3_3 = simpleScore(rankedIC3, upSet = sig3, centerScore = TRUE, knownDirection = FALSE)


projectScoreLandscape(plt1, scoredf_ic1, scoredf_ic2, annot = "cluster1")
projectScoreLandscape(plt2, scoredf_ic1, scoredf_ic3, annot = "cluster1")
projectScoreLandscape(plt3, scoredf_ic2_3, scoredf_ic3_3, annot = "cluster3")

alldf = data.frame("gs1" = scoredf1$TotalScore, "gs2" = scoredf2$TotalScore, "gs3" = scoredf3$TotalScore)



```


```{r plot}
library(ggplot2)

# The largest 65 gs1 scores are almost all cluster1
sum(tail(scoredf1_pdf[order(scoredf1_pdf$scoredf1.TotalScore),],65)$group != "iCluster1")

# -0.254947967 iCluster1
scoredf1_pdf = data.frame(files = rownames(scoredf1),scoredf1$TotalScore, group = dge_object$samples$group)

scoredf2_pdf = data.frame(files = rownames(scoredf2),scoredf2$TotalScore, group = dge_object$samples$group)


scoredf3_pdf = data.frame(files = rownames(scoredf3),scoredf3$TotalScore, group = dge_object$samples$group)

all = merge(merge(scoredf1_pdf, scoredf2_pdf, by.x = "files", by.y = "files"), scoredf3_pdf, by = "files")
```

```{r 3Dplot}
library(plotly)
all$group = as.factor(all$group)


p <- plot_ly(all, x = ~scoredf1.TotalScore, y = ~scoredf2.TotalScore, z = ~scoredf1.TotalScore, color = ~group, colors = c('red', 'blue', 'green')) %>%
  add_markers() %>%
  layout(scene = list(xaxis = list(title = 'iCGS1'),
                     yaxis = list(title = 'iCGS2'),
                     zaxis = list(title = 'iCGS3')))

# Create a shareable link to your chart
# Set up API credentials: https://plot.ly/r/getting-started
chart_link = api_create(p, filename="scatter3d-basic")
chart_link
```

```{r linearRegerssion}
allData = all[,c(2,4,6,7)]
allData$group = as.character(allData$group)
allData$group[allData$group=="iCluster1"] = 1
allData$group[allData$group=="iCluster2"] = 2
allData$group[allData$group=="iCluster3"] = 3

library(nnet)
regressionLm = multinom(group ~ ., data = allData, family = binomial(link="logit") )
summary(regressionLm)
predictions = predict(regressionLm, allData, type="class")

sum(predictions==allData$group)

errRate = sum(!predictions==allData$group)/length(allData$group)
```

